File: sdk/javascript/src/wallet.ts
import CryptoJS from 'crypto-js';
import { Buffer } from 'buffer';
import bs58 from 'bs58';
import { ShardXClient } from './client';
import { Transaction, TransactionRequest } from './models';
import { WalletError, ValidationError } from './errors';

/**
 * Wallet options
 */
export interface WalletOptions {
  /** ShardX client */
  client: ShardXClient;
  /** Private key (hex encoded) */
  privateKey?: string;
}

/**
 * ShardX wallet
 * 
 * Manages keys and signing transactions
 */
export class Wallet {
  private readonly client: ShardXClient;
  private readonly privateKey?: string;
  private readonly publicKey?: string;
  private readonly address?: string;

  /**
   * Create a new wallet
   * @param options Wallet options
   */
  constructor(options: WalletOptions) {
    this.client = options.client;
    
    if (options.privateKey) {
      this.privateKey = options.privateKey;
      this.publicKey = this.derivePublicKey(this.privateKey);
      this.address = this.deriveAddress(this.publicKey);
    }
  }

  /**
   * Create a new random wallet
   * @param client ShardX client
   * @returns New wallet
   */
  static createRandom(client: ShardXClient): Wallet {
    // Generate random private key
    const privateKey = CryptoJS.lib.WordArray.random(32).toString();
    
    return new Wallet({
      client,
      privateKey,
    });
  }

  /**
   * Create a wallet from mnemonic phrase
   * @param mnemonic Mnemonic phrase
   * @param client ShardX client
   * @returns Wallet
   */
  static fromMnemonic(mnemonic: string, client: ShardXClient): Wallet {
    // In a real implementation, this would use BIP39/BIP44
    // For simplicity, we'll just hash the mnemonic to get a private key
    const privateKey = CryptoJS.SHA256(mnemonic).toString();
    
    return new Wallet({
      client,
      privateKey,
    });
  }

  /**
   * Get wallet address
   * @returns Wallet address
   */
  getAddress(): string {
    if (!this.address) {
      throw new WalletError('Wallet not initialized with private key', 0, 'wallet_not_initialized');
    }
    
    return this.address;
  }

  /**
   * Get wallet public key
   * @returns Public key (hex encoded)
   */
  getPublicKey(): string {
    if (!this.publicKey) {
      throw new WalletError('Wallet not initialized with private key', 0, 'wallet_not_initialized');
    }
    
    return this.publicKey;
  }

  /**
   * Sign a message
   * @param message Message to sign
   * @returns Signature (hex encoded)
   */
  sign(message: string): string {
    if (!this.privateKey) {
      throw new WalletError('Wallet not initialized with private key', 0, 'wallet_not_initialized');
    }
    
    // In a real implementation, this would use proper cryptographic signing
    // For simplicity, we'll just use HMAC
    const signature = CryptoJS.HmacSHA256(message, this.privateKey).toString();
    
    return signature;
  }

  /**
   * Create and sign a transaction
   * @param to Recipient address
   * @param amount Amount to send
   * @param data Optional transaction data
   * @returns Signed transaction
   */
  async createTransaction(to: string, amount: string, data?: string): Promise<Transaction> {
    if (!this.privateKey || !this.address) {
      throw new WalletError('Wallet not initialized with private key', 0, 'wallet_not_initialized');
    }
    
    // Validate inputs
    if (!to) {
      throw new ValidationError('Recipient address is required');
    }
    
    if (!amount || parseFloat(amount) <= 0) {
      throw new ValidationError('Amount must be greater than 0');
    }
    
    // Create transaction request
    const txRequest: TransactionRequest = {
      from: this.address,
      to,
      amount,
      data,
      signature: '', // Will be filled below
    };
    
    // Create message to sign
    const message = `${txRequest.from}:${txRequest.to}:${txRequest.amount}:${txRequest.data || ''}`;
    
    // Sign the message
    txRequest.signature = this.sign(message);
    
    // Send transaction to the network
    return this.client.createTransaction(txRequest);
  }

  /**
   * Get account balance
   * @returns Account balance
   */
  async getBalance(): Promise<string> {
    if (!this.address) {
      throw new WalletError('Wallet not initialized with private key', 0, 'wallet_not_initialized');
    }
    
    const account = await this.client.getAccount(this.address);
    return account.balance;
  }

  /**
   * Get transaction history
   * @param limit Maximum number of transactions to return
   * @param offset Offset for pagination
   * @returns Array of transactions
   */
  async getTransactions(limit = 20, offset = 0): Promise<Transaction[]> {
    if (!this.address) {
      throw new WalletError('Wallet not initialized with private key', 0, 'wallet_not_initialized');
    }
    
    return this.client.getTransactionsByAddress(this.address, limit, offset);
  }

  /**
   * Derive public key from private key
   * @param privateKey Private key (hex encoded)
   * @returns Public key (hex encoded)
   */
  private derivePublicKey(privateKey: string): string {
    // In a real implementation, this would use proper cryptographic key derivation
    // For simplicity, we'll just hash the private key
    return CryptoJS.SHA256(privateKey).toString();
  }

  /**
   * Derive address from public key
   * @param publicKey Public key (hex encoded)
   * @returns Address
   */
  private deriveAddress(publicKey: string): string {
    // In a real implementation, this would use proper address derivation
    // For simplicity, we'll hash the public key and encode in base58
    const hash = CryptoJS.RIPEMD160(CryptoJS.SHA256(publicKey)).toString();
    const hashWithPrefix = '00' + hash; // Add version prefix
    
    // Convert to Buffer and encode in base58
    const buffer = Buffer.from(hashWithPrefix, 'hex');
    return bs58.encode(buffer);
  }
}

File: sdk/python/shardx_sdk/wallet.py
import os
import binascii
import hashlib
import hmac
import base58
from typing import List, Optional, Dict, Any
from Crypto.Random import get_random_bytes

from .client import ShardXClient
from .models import Transaction, TransactionRequest
from .errors import WalletError, ValidationError

class Wallet:
    """
    ShardX wallet
    
    Manages keys and signing transactions
    """
    
    def __init__(
        self,
        client: ShardXClient,
        private_key: Optional[str] = None
    ):
        """
        Initialize a new wallet
        
        Args:
            client: ShardX client
            private_key: Private key (hex encoded)
        """
        self.client = client
        
        if private_key:
            self.private_key = private_key
            self.public_key = self._derive_public_key(self.private_key)
            self.address = self._derive_address(self.public_key)
        else:
            self.private_key = None
            self.public_key = None
            self.address = None
    
    @classmethod
    def create_random(cls, client: ShardXClient) -> "Wallet":
        """
        Create a new random wallet
        
        Args:
            client: ShardX client
            
        Returns:
            New wallet
        """
        # Generate random private key
        private_key = binascii.hexlify(get_random_bytes(32)).decode('ascii')
        
        return cls(client, private_key=private_key)
    
    @classmethod
    def from_mnemonic(cls, mnemonic: str, client: ShardXClient) -> "Wallet":
        """
        Create a wallet from mnemonic phrase
        
        Args:
            mnemonic: Mnemonic phrase
            client: ShardX client
            
        Returns:
            Wallet
        """
        # In a real implementation, this would use BIP39/BIP44
        # For simplicity, we'll just hash the mnemonic to get a private key
        private_key = hashlib.sha256(mnemonic.encode()).hexdigest()
        
        return cls(client, private_key=private_key)
    
    def get_address(self) -> str:
        """
        Get wallet address
        
        Returns:
            Wallet address
            
        Raises:
            WalletError: If wallet is not initialized with private key
        """
        if not self.address:
            raise WalletError("Wallet not initialized with private key", code="wallet_not_initialized")
        
        return self.address
    
    def get_public_key(self) -> str:
        """
        Get wallet public key
        
        Returns:
            Public key (hex encoded)
            
        Raises:
            WalletError: If wallet is not initialized with private key
        """
        if not self.public_key:
            raise WalletError("Wallet not initialized with private key", code="wallet_not_initialized")
        
        return self.public_key
    
    def sign(self, message: str) -> str:
        """
        Sign a message
        
        Args:
            message: Message to sign
            
        Returns:
            Signature (hex encoded)
            
        Raises:
            WalletError: If wallet is not initialized with private key
        """
        if not self.private_key:
            raise WalletError("Wallet not initialized with private key", code="wallet_not_initialized")
        
        # In a real implementation, this would use proper cryptographic signing
        # For simplicity, we'll just use HMAC
        signature = hmac.new(
            bytes.fromhex(self.private_key),
            message.encode(),
            hashlib.sha256
        ).hexdigest()
        
        return signature
    
    async def create_transaction(
        self,
        to: str,
        amount: str,
        data: Optional[str] = None
    ) -> Transaction:
        """
        Create and sign a transaction
        
        Args:
            to: Recipient address
            amount: Amount to send
            data: Optional transaction data
            
        Returns:
            Signed transaction
            
        Raises:
            WalletError: If wallet is not initialized with private key
            ValidationError: If inputs are invalid
        """
        if not self.private_key or not self.address:
            raise WalletError("Wallet not initialized with private key", code="wallet_not_initialized")
        
        # Validate inputs
        if not to:
            raise ValidationError("Recipient address is required")
        
        try:
            if float(amount) <= 0:
                raise ValidationError("Amount must be greater than 0")
        except ValueError:
            raise ValidationError("Invalid amount")
        
        # Create transaction request
        tx_request = TransactionRequest(
            from_address=self.address,
            to=to,
            amount=amount,
            data=data,
            signature=""  # Will be filled below
        )
        
        # Create message to sign
        message = f"{tx_request.from_address}:{tx_request.to}:{tx_request.amount}:{tx_request.data or ''}"
        
        # Sign the message
        tx_request.signature = self.sign(message)
        
        # Send transaction to the network
        return self.client.create_transaction(tx_request)
    
    async def get_balance(self) -> str:
        """
        Get account balance
        
        Returns:
            Account balance
            
        Raises:
            WalletError: If wallet is not initialized with private key
        """
        if not self.address:
            raise WalletError("Wallet not initialized with private key", code="wallet_not_initialized")
        
        account = self.client.get_account(self.address)
        return account.balance
    
    async def get_transactions(
        self,
        limit: int = 20,
        offset: int = 0
    ) -> List[Transaction]:
        """
        Get transaction history
        
        Args:
            limit: Maximum number of transactions to return
            offset: Offset for pagination
            
        Returns:
            List of transactions
            
        Raises:
            WalletError: If wallet is not initialized with private key
        """
        if not self.address:
            raise WalletError("Wallet not initialized with private key", code="wallet_not_initialized")
        
        return self.client.get_transactions_by_address(self.address, limit, offset)
    
    def _derive_public_key(self, private_key: str) -> str:
        """
        Derive public key from private key
        
        Args:
            private_key: Private key (hex encoded)
            
        Returns:
            Public key (hex encoded)
        """
        # In a real implementation, this would use proper cryptographic key derivation
        # For simplicity, we'll just hash the private key
        return hashlib.sha256(bytes.fromhex(private_key)).hexdigest()
    
    def _derive_address(self, public_key: str) -> str:
        """
        Derive address from public key
        
        Args:
            public_key: Public key (hex encoded)
            
        Returns:
            Address
        """
        # In a real implementation, this would use proper address derivation
        # For simplicity, we'll hash the public key and encode in base58
        sha256_hash = hashlib.sha256(bytes.fromhex(public_key)).digest()
        ripemd160_hash = hashlib.new('ripemd160', sha256_hash).digest()
        
        # Add version prefix (0x00)
        versioned_hash = b'\x00' + ripemd160_hash
        
        # Encode in base58
        return base58.b58encode(versioned_hash).decode('ascii')